subcommand/get-consul-client-ca/command.go

```diff
@@ -0,0 +1,227 @@
+			"\"debug\", \"info\", \"warn\", and \"error\".")
+
+	c.help = flags.Usage(help, c.flags)
+}
+
+func (c *Command) Run(args []string) int {
+	c.once.Do(c.init)
+	if err := c.flags.Parse(args); err != nil {
+		return 1
+	}
+	if len(c.flags.Args()) > 0 {
+		c.UI.Error(fmt.Sprintf("Should have no non-flag arguments."))
+		return 1
+	}
+
+	if c.flagOutputFile == "" {
+		c.UI.Error(fmt.Sprintf("-output-file must be set"))
+		return 1
+	}
+
+	// create a logger
+	level := hclog.LevelFromString(c.flagLogLevel)
+	if level == hclog.NoLevel {
+		c.UI.Error(fmt.Sprintf("Unknown log level: %s", c.flagLogLevel))
+		return 1
+	}
+	logger := hclog.New(&hclog.LoggerOptions{
+		Level:  level,
+		Output: os.Stderr,
+	})
+
+	// create Consul client
+	consulClient, err := c.consulClient(logger)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error initializing Consul client: %s", err))
+		return 1
+	}
+
+	// Get the active CA root from Consul
+	// Wait until it gets a successful response
+	var activeRoot string
+	for activeRoot == "" {
+		caRoots, _, err := consulClient.Agent().ConnectCARoots(nil)
+		if err != nil {
+			logger.Info("Error retrieving CA roots from Consul", "err", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		activeRoot, err = getActiveRoot(caRoots)
+		if err != nil {
+			logger.Info("Could not get an active root", "err", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+	}
+
+	err = ioutil.WriteFile(c.flagOutputFile, []byte(activeRoot), 0644)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error writing CA file: %s", err))
+		return 1
+	}
+
+	c.UI.Info(fmt.Sprintf("Successfully written Consul client CA to: %s", c.flagOutputFile))
+	return 0
+}
+
+func (c *Command) consulClient(logger hclog.Logger) (*api.Client, error) {
+	cfg := api.DefaultConfig()
+
+	// First, check if the server address is a cloud auto-join string.
+	// If it is, discover server addresses through the cloud provider.
+	if strings.Contains(c.flagServerAddr, "provider=") {
+		disco, err := c.newDiscover()
+		if err != nil {
+			return nil, err
+		}
+		logger.Debug("using cloud auto-join with", c.flagServerAddr)
+		servers, err := disco.Addrs(c.flagServerAddr, logger.StandardLogger(&hclog.StandardLoggerOptions{
+			InferLevels: true,
+		}))
+		if err != nil {
+			return nil, err
+		}
+
+		// check if we discovered any servers
+		if len(servers) == 0 {
+			return nil, fmt.Errorf("could not discover any Consul servers with %q", c.flagServerAddr)
+		}
+
+		logger.Debug("discovered servers", strings.Join(servers, " "))
+
+		// Pick the first server from the list,
+		// ignoring the port since we need to use HTTP API
+		firstServer := strings.SplitN(servers[0], ":", 2)[0]
+		cfg.Address = fmt.Sprintf("%s:8501", firstServer)
+		cfg.Scheme = "https"
+	} else {
+		// check if the server URL is missing a port
+		host := strings.TrimPrefix(c.flagServerAddr, "https://")
+		host = strings.TrimPrefix(c.flagServerAddr, "http://")
+		parts := strings.SplitN(host, ":", 2)
```

Comment:
Yeah I tried using `url.Parse`. The problem with that function is that it's [ambiguous](https://golang.org/pkg/net/url/#Parse) if the scheme is missing. 

I don't think we should return error because my general intent with this was to assume secure defaults when not enough information about the server is provided (like in the case of cloud auto-join), but if someone really wants/needs to use HTTP, I think it should be allowed.

Reply:
Also, it simplifies tests quite a bit because we don't have to generate certs for all test agents we're spinning up 😄